_source/_posts/2020-04-29-master-filestream-in-csharp-aspnetcore.md

```diff
@@ -0,0 +1,321 @@
+                this.Response.Headers.Add("Content-Disposition", "attachment; filename=\"filename.json\"");
+                this.Response.StatusCode = 200;
+                await fileStream.CopyToAsync(this.Response.Body);
+            }
+        }
+    }
+}
+```
+
+Let's break down what the code above is doing. The first thing we want to do is make sure no one that is logged in can access our controller. As Okta is already set up, all we need to do is add the **Authorize** decorator to the controller class.
+
+```csharp
+[Authorize]
+public class FileController : ControllerBase
+```
+
+Next, since the AWS SDK took care of loading our s3 client into our services, we can access it via dependency injection.
+
+```csharp
+IAmazonS3 _s3Client;
+public FileController(IAmazonS3 s3Client)
+{
+    _s3Client = s3Client;
+}
+```
+
+Okta gives us access to any information about our user that we ask for. This lets us have pretty robust logic for authorization. In my case, I'm filtering by email, but this could be by domain, role, age, or any other arbitrary claim that helps you determine authorization.
+
+```csharp
+var emailClaim = User.Claims.FirstOrDefault(x => x.Type == "email");
+var email = emailClaim.Value;
+
+if (email == "chase@cupcakefactory.net")
+```
+
+Lastly, because of .NET Core, we have direct access to the response body stream and can stream our bytes directly to the output without having to download the whole content first. We can also do any processing of the data before writing to the output stream the way we did in our first example.
+
+```csharp
+using (var fileStream = await _s3Client.GetObjectStreamAsync("your-bucket", "somefile", null))
+{
+    //Whatever your content type is
+    this.Response.ContentType = "application/json";
+    this.Response.Headers.Add("Content-Disposition", "attachment; filename=\"myfile.json\"");
+    this.Response.StatusCode = 200;
+    await fileStream.CopyToAsync(this.Response.Body);
+}
+```
+
+## Final thoughts
+
+I'm going to share an anecdote with you that explains a particular time this was very useful to me. I was working with a company that sent tens of millions of emails a month. Some of these emails were very large and we were hitting some massive bottlenecks with our infrastructure. We were able to write our own SMTP client to more efficiently chuck the data and manage memory, allowing us to go from emails being backed up for hours to delivering emails in milliseconds - still using the same hardware. It's never a great idea to optimize early, but understanding the low-level protocols can be a lifesaver if you are dealing with especially large volumes of data.
+
+Today we've learned how to break down large tasks into small chunks and how to build a secure streaming service. I hope this has given you a starting place for working with some lower-level  parts of the .NET framework. They can seem scary but, if you just think about them as reading and writing data, it gets a lot less daunting.
```

Comment:
This section is missing a source code repo? If there isn't one, that's fine. However, if developers have comments and say they're unable to make it work -- it's nice to have a repo to try it out with and see if they're right.

Reply:
I agree that it would be ideal. I'll add that to the template for the contractor.